ports/rp2/rp2_pio.c

```diff
@@ -462,21 +462,32 @@ STATIC mp_obj_t rp2_state_machine_init_helper(const rp2_state_machine_obj_t *sel
     }
 
     // Compute the clock divider.
-    float div;
+    uint16_t clkdiv_int;
+    uint8_t clkdiv_frac;
     if (args[ARG_freq].u_int < 0) {
-        div = 1;
+        // Default: run at CPU frequency.
+        clkdiv_int = 1;
+        clkdiv_frac = 0;
     } else if (args[ARG_freq].u_int == 0) {
-        div = 0;
+        // Special case of 0: set clkdiv to 0.
+        clkdiv_int = 0;
+        clkdiv_frac = 0;
     } else {
-        div = (float)clock_get_hz(clk_sys) / (float)args[ARG_freq].u_int;
+        // Frequency given in Hz, compute clkdiv from it.
+        uint64_t div = (uint64_t)clock_get_hz(clk_sys) * 256ULL / (uint64_t)args[ARG_freq].u_int;
+        if (!(div >= 1 * 256 && div <= 65535 * 256)) {
+            mp_raise_ValueError(MP_ERROR_TEXT("freq out of range"));
```

Comment:
Would it be helpful for this error message to give the min and max allowed SM freq for the current CPU frequency? :shrug:

Reply:
It's a bit of work to do that, but could be useful (at least saves people reading the datasheet...!).

But since the bounds would be something like 1907.349 to 125e6, it makes me think that the `freq` argument should probably accept floats...

Reply:
Yeah, I guess there's no reason not to expose non-integer frequencies.

Reply:
In the interests of getting this fix in, let's leave it for another PR to expose non-integer frequencies.